### PR TITLE
[DO NOT MERGE] 🐛 fix  race condition in amp-geo

### LINKED ISSUE #14841: Race condition in services provided by extensions.
There looks to be a race in the service promise for `amp-geo`.   Repro steps:

Load a page with amp geo
Clear the browser cache
Load `v0.js, v0.css`, etc but not `amp-geo` etc in another browser window to prime cache.
Reload the `amp-geo` page.

Result is a console error `To use variable  AMP_GEO , amp-geo should be configured`

### PR DESCRIPTION
**Pull request by jpettitt**
Fixes #14841 

Prevent race condition with extension services returning null due to loading order issues.

Fix make getElmentServiceIfAvailableForDoc and getElementServicePromiseOrNull wait for doc ready rather than waiting for body.

Move amp-geo service promise creation from the `buildCallBack()` to the `constructor()`.

### CODE REVIEW

File: extensions/amp-geo/0.1/amp-geo.js [hunk 1]
```diff
@@ -78,14 +79,23 @@ export class AmpGeo extends AMP.BaseElement {
...
     super(element);
+    const self = this;
 
     /** @private {number} */
     this.mode_ = mode.GEO_HOT_PATCH;
     /** @private {string} */
     this.country_ = 'unknown';
     /** @private {Array<string>} */
     this.matchedGroups_ = [];
+    /** @private {?Promise} */
```
Comment (cramforce) [inline extensions/amp-geo/0.1/amp-geo.js]: `!Function`

Comment (jridgewell) [reply extensions/amp-geo/0.1/amp-geo.js]: `function()`, `Function` was discouraged.

File: extensions/amp-geo/0.1/amp-geo.js [hunk 2]
```diff
@@ -78,14 +79,23 @@ export class AmpGeo extends AMP.BaseElement {
...
     super(element);
+    const self = this;
 
     /** @private {number} */
     this.mode_ = mode.GEO_HOT_PATCH;
     /** @private {string} */
     this.country_ = 'unknown';
     /** @private {Array<string>} */
     this.matchedGroups_ = [];
+    /** @private {?function()} */
```
Comment (cramforce) [inline extensions/amp-geo/0.1/amp-geo.js]: Sorry, actual correct type is `?function(!Object<string, (string|Array<string>)>)`

Comment (jpettitt) [reply extensions/amp-geo/0.1/amp-geo.js]: The accompanying change to `element-service` should make it work per @jridgewell by waiting until doc ready.  Will fix the type def.

**Comment by jridgewell** [comment]
We can define the service promise in the top level of the module, that’d be fine. It would also mean we could continue to use the faster `whenBodyAvailable`, which it a win too.

If we take that approach, we should also follow up with PRs for the other element-provided services.

**Comment by jpettitt** [comment]
Adding `[do not merge]` There is still a race condition in the extreme case where the extension js file doesn't get loaded until after document ready.  This hits all element services no matter how they create the service because if their JS hasn't yet loaded the runtime doesn't know about the service.

**Comment by jpettitt** [comment]
Closing this because we're going a different route. See #14841

**Pull request closed without merging**